tests/README.md

```diff
@@ -0,0 +1,515 @@
+
+
+### Adding an Attestation Test
+
+Attestations can't happen in the same block as the one about which they are attesting, or in a block that is
+after the block is finalized. This is specified as part of the specs, in the `process_attestation` function
+(which is created from the spec by the `make pyspec` command you ran earlier). Here is the relevant code
+fragment:
+
+
+```python
+def process_attestation(state: BeaconState, attestation: Attestation) -> None:
+    data = attestation.data
+    assert data.target.epoch in (get_previous_epoch(state), get_current_epoch(state))
+    assert data.target.epoch == compute_epoch_at_slot(data.slot)
+    assert data.slot + MIN_ATTESTATION_INCLUSION_DELAY <= state.slot <= data.slot + SLOTS_PER_EPOCH
+    ...
+```
+
+In the last line you can see two conditions being asserted:
+
+1. `data.slot + MIN_ATTESTATION_INCLUSION_DELAY <= state.slot` which verifies that the attestation doesn't
+   arrive too early.
+1. `state.slot <= data.slot + SLOTS_PER_EPOCH` which verifies that the attestation doesn't
+   arrive too late.
+   
+This is how the consensus layer tests deal with edge cases, by asserting the conditions required for the
+values to be legitimate. In the case of these particular conditions, they are tested 
+[here](https://github.com/ethereum/consensus-specs/blob/926e5a3d722df973b9a12f12c015783de35cafa9/tests/core/pyspec/eth2spec/test/phase0/block_processing/test_process_attestation.py#L87-L104).
+One test checks what happens if the attestation is too early, and another if it is too late.
+
+However, it is not enough to ensure we reject invalid blocks. It is also necessary to ensure we accept all valid blocks. You saw earlier
+a test (`test_success`) that tested that being `MIN_ATTESTATION_INCLUSION_DELAY` after the data for which we attest is enough. 
+Now we'll write a similar test that verifies that being `SLOTS_PER_EPOCH` away is still valid. To do this, we modify the 
+`test_after_epoch_slots` function. We need two changes:
+
+1. Call `transition_to_slot_via_block` with one less slot to advance
+1. Don't tell `run_attestation_processing` to return an empty post state.
+
+The modified function is:
+
+```python
+@with_all_phases
+@spec_state_test
+def test_almost_after_epoch_slots(spec, state):
+    attestation = get_valid_attestation(spec, state, signed=True)
+
+    # increment to latest inclusion slot (not beyond it)
+    transition_to_slot_via_block(spec, state, state.slot + spec.SLOTS_PER_EPOCH)
+
+    yield from run_attestation_processing(spec, state, attestation)
+```    
+
+Add this function to the file `consensus-specs/tests/core/pyspec/eth2spec/test/phase0/block_processing/test_process_attestation.py`,
+and run the test:
+
+```sh
+cd ~/consensus-specs
+. venv/bin/activate
+cd tests/core/pyspec
+python -m pytest -k almost_after --fork Merge eth2spec
+```
+
+You should see it ran successfully (although you might get a warning, you can ignore it)
+
+<!--
+
+https://ethos.dev/beacon-chain/
```

Comment:
This is an older website, and maybe not so relevant for testing. We can aggregate starter resources in another place.